File: src/lib/backlogAuth.ts
export type BacklogProfile = {
  id: number;
  name: string;
  userId?: string;
  mailAddress?: string;
  roleType?: number;
  lang?: string;
  nickName?: string;
};

export type BacklogAuthResult =
  | { status: "success"; profile: BacklogProfile; baseUrl: string }
  | { status: "error"; error: string; baseUrl: string };

export type BacklogAuthParams = {
  host: string;
  apiKey: string;
  fetchFn?: typeof fetch;
};

export function buildBacklogBaseUrl(host: string): string {
  const trimmed = host.trim();
  if (!trimmed) return "";
  if (trimmed.startsWith("http://") || trimmed.startsWith("https://")) {
    return trimmed.replace(/\/+$/, "");
  }
  return `https://${trimmed.replace(/\/+$/, "")}`;
}

export async function fetchBacklogProfile(
  params: BacklogAuthParams
): Promise<BacklogAuthResult> {
  const baseUrl = buildBacklogBaseUrl(params.host);
  if (!baseUrl) {
    return {
      status: "error",
      error: "Vui lòng nhập Backlog host (ví dụ: example.backlog.com)",
      baseUrl,
    };
  }
  if (!params.apiKey.trim()) {
    return {
      status: "error",
      error: "Vui lòng nhập API key Backlog",
      baseUrl,
    };
  }

  const fetchImpl = params.fetchFn ?? fetch;
  const url = `${baseUrl}/api/v2/users/myself?apiKey=${encodeURIComponent(
    params.apiKey.trim()
  )}`;

  try {
    const res = await fetchImpl(url, { method: "GET" });
    if (!res.ok) {
      const text = await res.text();
      return {
        status: "error",
        error:
          `Không lấy được profile (${res.status}). ` +
          (text || "Vui lòng kiểm tra API key/host."),
        baseUrl,
      };
    }

    const data = (await res.json()) as BacklogProfile;
    return { status: "success", profile: data, baseUrl };
  } catch (e: any) {
    return {
      status: "error",
      error: `Lỗi kết nối: ${String(e)}`,
      baseUrl,
    };
  }
}

export type BacklogOAuthToken = {
  access_token: string;
  token_type: "Bearer";
  expires_in: number;
  refresh_token: string;
};

export type BacklogOAuthTokenResult =
  | { status: "success"; token: BacklogOAuthToken; baseUrl: string }
  | { status: "error"; error: string; baseUrl: string };

export type BacklogOAuthAuthorizeParams = {
  host: string;
  clientId: string;
  redirectUri: string;
  state: string;
};

export type BacklogOAuthTokenParams = {
  host: string;
  clientId: string;
  clientSecret: string;
  redirectUri: string;
  code: string;
  fetchFn?: typeof fetch;
};

export type BacklogOAuthRefreshParams = {
  host: string;
  clientId: string;
  clientSecret: string;
  refreshToken: string;
  fetchFn?: typeof fetch;
};

export function buildBacklogOAuthAuthorizeUrl(
  params: BacklogOAuthAuthorizeParams
): string {
  const baseUrl = buildBacklogBaseUrl(params.host);
  if (!baseUrl) return "";
  const search = new URLSearchParams({
    response_type: "code",
    client_id: params.clientId,
    redirect_uri: params.redirectUri,
    state: params.state,
  });
  return `${baseUrl}/OAuth2AccessRequest.action?${search.toString()}`;
}

export async function exchangeBacklogOAuthToken(
  params: BacklogOAuthTokenParams
): Promise<BacklogOAuthTokenResult> {
  const baseUrl = buildBacklogBaseUrl(params.host);
  if (!baseUrl) {
    return {
      status: "error",
      error: "Vui lòng nhập Backlog host (ví dụ: example.backlog.com)",
      baseUrl,
    };
  }
  if (!params.clientId.trim() || !params.clientSecret.trim()) {
    return {
      status: "error",
      error: "Vui lòng nhập client_id và client_secret",
      baseUrl,
    };
  }
  if (!params.redirectUri.trim()) {
    return {
      status: "error",
      error: "Vui lòng nhập redirect_uri",
      baseUrl,
    };
  }
  const fetchImpl = params.fetchFn ?? fetch;
  const body = new URLSearchParams({
    grant_type: "authorization_code",
    code: params.code,
    redirect_uri: params.redirectUri,
    client_id: params.clientId,
    client_secret: params.clientSecret,
  });
  try {
    const res = await fetchImpl(`${baseUrl}/api/v2/oauth2/token`, {
      method: "POST",
      headers: {
        "Content-Type": "application/x-www-form-urlencoded",
      },
      body,
    });
    if (!res.ok) {
      const text = await res.text();
      return {
        status: "error",
        error:
          `Không lấy được token (${res.status}). ` +
          (text || "Vui lòng kiểm tra client_id/secret/redirect_uri."),
        baseUrl,
      };
    }
    const token = (await res.json()) as BacklogOAuthToken;
    return { status: "success", token, baseUrl };
  } catch (e: any) {
    return {
      status: "error",
      error: `Lỗi kết nối: ${String(e)}`,
      baseUrl,
    };
  }
}

export async function refreshBacklogOAuthToken(
  params: BacklogOAuthRefreshParams
): Promise<BacklogOAuthTokenResult> {
  const baseUrl = buildBacklogBaseUrl(params.host);
  if (!baseUrl) {
    return {
      status: "error",
      error: "Vui lòng nhập Backlog host (ví dụ: example.backlog.com)",
      baseUrl,
    };
  }
  if (!params.clientId.trim() || !params.clientSecret.trim()) {
    return {
      status: "error",
      error: "Vui lòng nhập client_id và client_secret",
      baseUrl,
    };
  }
  if (!params.refreshToken.trim()) {
    return {
      status: "error",
      error: "Vui lòng nhập refresh_token",
      baseUrl,
    };
  }
  const fetchImpl = params.fetchFn ?? fetch;
  const body = new URLSearchParams({
    grant_type: "refresh_token",
    client_id: params.clientId,
    client_secret: params.clientSecret,
    refresh_token: params.refreshToken,
  });
  try {
    const res = await fetchImpl(`${baseUrl}/api/v2/oauth2/token`, {
      method: "POST",
      headers: {
        "Content-Type": "application/x-www-form-urlencoded",
      },
      body,
    });
    if (!res.ok) {
      const text = await res.text();
      return {
        status: "error",
        error:
          `Không refresh được token (${res.status}). ` +
          (text || "Vui lòng kiểm tra refresh_token."),
        baseUrl,
      };
    }
    const token = (await res.json()) as BacklogOAuthToken;
    return { status: "success", token, baseUrl };
  } catch (e: any) {
    return {
      status: "error",
      error: `Lỗi kết nối: ${String(e)}`,
      baseUrl,
    };
  }
}

export async function fetchBacklogProfileWithToken(
  params: { host: string; accessToken: string; fetchFn?: typeof fetch }
): Promise<BacklogAuthResult> {
  const baseUrl = buildBacklogBaseUrl(params.host);
  if (!baseUrl) {
    return {
      status: "error",
      error: "Vui lòng nhập Backlog host (ví dụ: example.backlog.com)",
      baseUrl,
    };
  }
  if (!params.accessToken.trim()) {
    return {
      status: "error",
      error: "Vui lòng nhập access_token",
      baseUrl,
    };
  }
  const fetchImpl = params.fetchFn ?? fetch;
  try {
    const res = await fetchImpl(`${baseUrl}/api/v2/users/myself`, {
      method: "GET",
      headers: {
        Authorization: `Bearer ${params.accessToken}`,
      },
    });
    if (!res.ok) {
      const text = await res.text();
      return {
        status: "error",
        error:
          `Không lấy được profile (${res.status}). ` +
          (text || "Vui lòng kiểm tra access_token."),
        baseUrl,
      };
    }
    const data = (await res.json()) as BacklogProfile;
    return { status: "success", profile: data, baseUrl };
  } catch (e: any) {
    return {
      status: "error",
      error: `Lỗi kết nối: ${String(e)}`,
      baseUrl,
    };
  }
}

File: src-tauri/src/backlog_auth.rs
use serde::{Deserialize, Serialize};
use std::env;
use urlencoding::encode;

#[derive(Serialize, Deserialize)]
pub struct BacklogOAuthToken {
    pub access_token: String,
    pub token_type: String,
    pub expires_in: u64,
    pub refresh_token: String,
    #[serde(default)]
    pub host: String, // Added to sync with frontend, not in Backlog API response
}

#[derive(Serialize, Deserialize)]
pub struct BacklogConfig {
    pub host: String,
    pub client_id: String,
    pub redirect_uri: String,
}

fn normalize_base_url(host: &str) -> Result<String, String> {
    let trimmed = host.trim();
    if trimmed.is_empty() {
        return Err("BACKLOG_HOST is not configured".to_string());
    }
    let mut base = if trimmed.starts_with("http://") || trimmed.starts_with("https://") {
        trimmed.to_string()
    } else {
        format!("https://{}", trimmed)
    };
    while base.ends_with('/') {
        base.pop();
    }
    Ok(base)
}

fn get_backlog_env(key: &str) -> String {
    match key {
        "BACKLOG_HOST" => env!("BACKLOG_HOST").to_string(),
        "BACKLOG_CLIENT_ID" => env!("BACKLOG_CLIENT_ID").to_string(),
        "BACKLOG_CLIENT_SECRET" => env!("BACKLOG_CLIENT_SECRET").to_string(),
        "BACKLOG_REDIRECT_URI" => env!("BACKLOG_REDIRECT_URI").to_string(),
        _ => "".to_string(),
    }
}

#[tauri::command]
pub fn get_backlog_config() -> Result<BacklogConfig, String> {
    let host = get_backlog_env("BACKLOG_HOST");
    let client_id = get_backlog_env("BACKLOG_CLIENT_ID");
    let redirect_uri = get_backlog_env("BACKLOG_REDIRECT_URI");



    Ok(BacklogConfig {
        host,
        client_id,
        redirect_uri,
    })
}

#[tauri::command]
pub fn get_backlog_auth_url(state: String) -> Result<String, String> {
    let host = get_backlog_env("BACKLOG_HOST");
    let client_id = get_backlog_env("BACKLOG_CLIENT_ID");
    let redirect_uri = get_backlog_env("BACKLOG_REDIRECT_URI");



    if host.is_empty() {
        return Err("Missing BACKLOG_HOST in .env".to_string());
    }
    if client_id.is_empty() {
        return Err("Missing BACKLOG_CLIENT_ID in .env".to_string());
    }
    if redirect_uri.is_empty() {
        return Err("Missing BACKLOG_REDIRECT_URI in .env".to_string());
    }

    let base_url = normalize_base_url(&host)?;
    
    let url = format!(
        "{}/OAuth2AccessRequest.action?response_type=code&client_id={}&redirect_uri={}&state={}",
        base_url,
        encode(&client_id),
        encode(&redirect_uri),
        encode(&state)
    );

    Ok(url)
}

#[tauri::command]
pub fn backlog_oauth_exchange(code: String) -> Result<BacklogOAuthToken, String> {
    let host = get_backlog_env("BACKLOG_HOST");
    let client_id = get_backlog_env("BACKLOG_CLIENT_ID");
    let client_secret = get_backlog_env("BACKLOG_CLIENT_SECRET");
    let redirect_uri = get_backlog_env("BACKLOG_REDIRECT_URI");



    if host.is_empty() {
        return Err("Missing BACKLOG_HOST".to_string());
    }
    if client_id.is_empty() {
        return Err("Missing BACKLOG_CLIENT_ID".to_string());
    }
    if client_secret.is_empty() {
        return Err("Missing BACKLOG_CLIENT_SECRET".to_string());
    }
    if redirect_uri.is_empty() {
        return Err("Missing BACKLOG_REDIRECT_URI".to_string());
    }

    let base_url = normalize_base_url(&host)?;
    let params = [
        ("grant_type", "authorization_code"),
        ("code", code.as_str()),
        ("redirect_uri", redirect_uri.as_str()),
        ("client_id", client_id.as_str()),
        ("client_secret", client_secret.as_str()),
    ];

    let client = reqwest::blocking::Client::new();
    let res = client
        .post(format!("{}/api/v2/oauth2/token", base_url))
        .header("User-Agent", "BSNiSync/1.0.0")
        .header("Content-Type", "application/x-www-form-urlencoded")
        .form(&params)
        .send()
        .map_err(|e| format!("API connection error: {}", e))?;

    let status = res.status();
    let body = res.text().unwrap_or_default();

    println!("[backlog-auth] Token exchange response status: {}", status);
    if !status.is_success() {
        println!("[backlog-auth] Error body: {}", body);
        return Err(format!("Token exchange error ({}): {}", status, body));
    }

    let mut token: BacklogOAuthToken = serde_json::from_str(&body)
        .map_err(|e| format!("JSON token parse error: {} | Content: {}", e, body))?;
    
    token.host = host; // Populate host for frontend sync

    println!("[backlog-auth] Token exchange success.");
    Ok(token)
}

#[tauri::command]
pub fn backlog_oauth_refresh(refresh_token: String) -> Result<BacklogOAuthToken, String> {
    let host = get_backlog_env("BACKLOG_HOST");
    let client_id = get_backlog_env("BACKLOG_CLIENT_ID");
    let client_secret = get_backlog_env("BACKLOG_CLIENT_SECRET");



    if host.is_empty() {
        return Err("Missing BACKLOG_HOST".to_string());
    }
    if client_id.is_empty() {
        return Err("Missing BACKLOG_CLIENT_ID".to_string());
    }
    if client_secret.is_empty() {
        return Err("Missing BACKLOG_CLIENT_SECRET".to_string());
    }

    let base_url = normalize_base_url(&host)?;
    let params = [
        ("grant_type", "refresh_token"),
        ("client_id", client_id.as_str()),
        ("client_secret", client_secret.as_str()),
        ("refresh_token", refresh_token.as_str()),
    ];

    let client = reqwest::blocking::Client::new();
    let res = client
        .post(format!("{}/api/v2/oauth2/token", base_url))
        .header("User-Agent", "BSNiSync/1.0.0")
        .header("Content-Type", "application/x-www-form-urlencoded")
        .form(&params)
        .send()
        .map_err(|e| format!("API connection error: {}", e))?;

    let status = res.status();
    let body = res.text().unwrap_or_default();

    println!("[backlog-auth] Token refresh response status: {}", status);
    if !status.is_success() {
        println!("[backlog-auth] Error body: {}", body);
        return Err(format!("Token refresh error ({}): {}", status, body));
    }

    let mut token: BacklogOAuthToken = serde_json::from_str(&body)
        .map_err(|e| format!("JSON token parse error: {} | Content: {}", e, body))?;
    
    token.host = host; // Populate host for frontend sync

    println!("[backlog-auth] Token refresh success.");
    Ok(token)
}
